components/permission-controller/controllers/namespace_controller.go

```diff
@@ -0,0 +1,142 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1beta1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	RolebindingName = "namespace-admin"
+	RoleRefKind     = "ClusterRole"
+	RoleRefName     = "kyma-admin"
+
+	subjectKindGroup  = "Group"
+	subjectKindUser   = "User"
+	SubjectStaticUser = "admin@kyma.cx"
+)
+
+// NamespaceReconciler reconciles a Namespace object
+type NamespaceReconciler struct {
+	client.Client
+	Log                logr.Logger
+	Scheme             *runtime.Scheme
+	ExcludedNamespaces []string
+	SubjectGroups      []string
+	UseStaticConnector bool
+}
+
+// Reconcile performs the reconciling for a single request object that can be used to fetch the namespace it represents from the cache
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=namespaces/status,verbs=get;update;patch;watch
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=rolebindings,verbs=create;update;patch;delete
+func (r *NamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("namespace", req.NamespacedName)
+
+	var namespace corev1.Namespace
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &namespace); err != nil {
+
+		if apierrs.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	if !namespace.ObjectMeta.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	if r.isProtected(namespace.Name) {
+		log.Info(fmt.Sprintf("%s is a system namespace. Skipping...", namespace.Name))
+		return ctrl.Result{}, nil
+	}
+
+	rb := &rbac.RoleBinding{
```

Comment:
Extract this to a dedicated function, it will be easier to read.

Reply:
see "address review comments 1" commit